Rust/src/sampled_tree.rs

```diff
@@ -0,0 +1,267 @@
+            AddResult::MassIncreased(key) => key,
+        };
+
+        match self.sampler.sample(point_key, sequence_index) {
+            SamplerResult::Accepted(evicted) => match evicted {
+                Some(evicted) => {
+                    // TODO: can we satisfy the borrow checker so that we can
+                    // perform the delete without needing to clone the point?
+                    let evicted_point = {
+                        let point_store = self.point_store.borrow();
+                        point_store.get(*evicted.value()).unwrap().clone()
+                    };
+                    self.tree.delete_point(&evicted_point);
+                }
+                None => ()
+            },
+            SamplerResult::Ignored => { self.tree.delete_point(&point); }
+        }
+    }
+
+    /// Traverse the tree with a given query point as input.
+    ///
+    /// Returns an iterator on the nodes of the tree. The iterator begins at the
+    /// tree's root node and follows the branch to the leaf node containing the
+    /// point nearest to the query point in L1 norm.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use random_cut_forest::{Node, SampledTree};
+    /// let mut tree: SampledTree<f32> = SampledTree::new(32, 0.01);
+    /// tree.update(vec![0.0, 0.0], 0);
+    /// tree.update(vec![1.0, 1.0], 1);
+    ///
+    /// // provide a query point that will always be to the left of any cut
+    /// // on the two points inserted above, no matter the cut location
+    /// let query = vec![-2.0, -2.0];
+    /// let nodes: Vec<&Node<f32>> = tree.traverse(&query).collect();
+    /// assert_eq!(nodes.len(), 2);
+    ///
+    /// // verify the last point of the tree is equal to the left-most point
+    /// if let Node::Leaf(leaf) = nodes[1] {
+    ///     let point_key = leaf.point();
+    ///     let point_store = tree.borrow_point_store();
+    ///     let point = point_store.get(point_key).unwrap();
+    ///     assert_eq!(point, &vec![0.0, 0.0]);
+    /// } else {
+    ///     panic!("Last node in traversal should be a leaf!")
+    /// }
+    /// ```
+    ///
+    pub fn traverse<'a>(&'a self, point: &'a Vec<T>) -> NodeTraverser<'a, T> {
+        self.tree.traverse(point)
+    }
+
+    /// Returns the sample size of the sampled tree.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use random_cut_forest::SampledTree;
+    ///
+    /// let tree: SampledTree<f32> = SampledTree::new(256, 0.001);
+    /// assert_eq!(tree.sample_size(), 256);
+    /// ```
+    pub fn sample_size(&self) -> usize { self.sampler.capacity() }
+
+    /// Returns the decay factor of the random sampler.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use random_cut_forest::SampledTree;
+    ///
+    /// let tree: SampledTree<f32> = SampledTree::new(256, 0.001);
+    /// assert_eq!(tree.lambda(), 0.001);
+    /// ```
+    pub fn lambda(&self) -> f64 { self.sampler.lambda() }
```

Comment:
I support this. In the Python version that I already have written I refer to it as `decay_factor` since `lambda` is a keyword. May as well change it to `time_decay` here and there.